### ISSUE TITLE
Deteriorated gui of files replace/merge window (xfce - fedora 40)

### ISSUE THREAD

**Issue by TNTBOMBOM** [QubesOS/qubes-issues#9374]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### Qubes OS release

4.2.2 (R.2)

### Steps to reproduce

- Create an AppVM based on fedora-40
- Open it and try to make duplicated files on different paths lets say /Download /Pictures
- Now try to copy/paste from one path to another
- Result should be shown similar to the screenshot
- Options are missing from the bottom of the window

### Screenshot

![mergebad](https://github.com/user-attachments/assets/92b310d4-7cae-4138-882c-efec75d560d5)

**Comment by alimirjamali** [comment]
I can reproduce it. But that is not XFCE's Thunar. It is actually GNOME's File Manager.

The symptom could be reproduced on Fedora 40 (GNOME) template by expanding GNOME Applications. They redraws correctly on shrink. Here is a snapshot of GNOME Calculator after expanding the window.

![gnome-calc](https://github.com/user-attachments/assets/9c0574da-0ae5-4c1e-9310-422da97a8b1e)

**Comment by DemiMarie** [comment]
Does unsetting `GSK_RENDERER` solve this problem?  The obvious reason for this is that Qubes OS forces the deprecated Cairo renderer to avoid GTK trying to use software-emulated OpenGL or Vulkan.

**Comment by alimirjamali** [comment]
> Does unsetting `GSK_RENDERER` solve this problem? The obvious reason for this is that Qubes OS forces the deprecated Cairo renderer to avoid GTK trying to use software-emulated OpenGL or Vulkan.

Actually `GSK_RENDERER` environment variable is unset by default. Here is the behaviour with different renderers (on gtk4.14.4-2):

unset: The exact symptom
boradway: N/A
cairo: Works well without issue (impressive for a software which is unmaintained for over a decade)
![cairo](https://github.com/user-attachments/assets/eef2a442-c6de-464d-b290-80d4cc8188fd)

opengl, gl, ngl: The exact symptom. Sometimes incomplete render at window bottom. Sometimes window does not warp-up correctly. Below is a screenshot of the symptom and related messages
![gl](https://github.com/user-attachments/assets/f1223a6e-6832-4d15-b91e-6d1fabd03759)

vulkan: Segfaults after few window resizes. Below is the screenshot of messages:
![vulkan](https://github.com/user-attachments/assets/fc5d594f-5789-4b22-851b-e104ccb74be6)

**Comment by alimirjamali** [comment]
The fix for #9268 by @ben-grande should unintentionally fix this as well. If the deprecated (but working) cairo renderer is going to be the default.

This is the patch which was merged few hours ago: https://github.com/QubesOS/qubes-gui-agent-linux/pull/214/files

**Comment by kennethrrosen** [comment]
I'm still experiencing this in vms running Gnome. I applied fix from https://github.com/QubesOS/qubes-issues/issues/9268 to the vm in question (those changes had not appeared in a Fedora-based standalone) but the windows are still cut-off. I'm using the latest kernel and have changed vcpus and maxmem to no avail.

**Comment by alimirjamali** [comment]
> I'm still experiencing this in vms running Gnome. I applied fix from #9268 to the vm in question (those changes had not appeared in a Fedora-based standalone) but the windows are still cut-off. I'm using the latest kernel and have changed vcpus and maxmem to no avail.

An additional patch is necessary. it might be still in main branch (not even in testing release). It is discussed [here](https://github.com/QubesOS/qubes-issues/issues/9389#issuecomment-2299160633).

**Comment by kennethrrosen** [comment]
cairo software rendering is indeed applied to all in /qubes-sysinit.sh (as patched in https://github.com/QubesOS/qubes-issues/issues/9389) however the problem persists.

**Comment by kennethrrosen** [comment]
Is there a way to change a VM's DE from Gnome to XFCE, or vise versa? I asked in the forum some while ago, but received no response.

**Comment by kennethrrosen** [comment]
> > Is there a way to change a VM's DE from Gnome to XFCE, or vise versa?
> 
> You can use [Xfce templates](https://www.qubes-os.org/doc/templates/xfce/).

Yes, of course; but in the case of a StandaloneVM?

**Comment by alimirjamali** [comment]
> cairo software rendering is indeed applied to all in /qubes-sysinit.sh (as patched in #9389) however the problem persists.

Interesting. Are you certain that the environment variable for `GSK_RENDERER` is set to `cairo` in the faulty qubes? Would you please double check?

**Comment by kennethrrosen** [comment]
> Interesting. Are you certain that the environment variable for `GSK_RENDERER` is set to `cairo` in the faulty qubes? Would you please double check?

Yes, qubes-gui.sh and qubes-gui.csh both have been updated per https://github.com/ben-grande/qubes-gui-agent-linux/commit/23c1eefa89a5c9c299c6f8e66a13b535e7e470d5

**Comment by alimirjamali** [comment]
> Yes, qubes-gui.sh and qubes-gui.csh both have been updated per [ben-grande/qubes-gui-agent-linux@23c1eef](https://github.com/ben-grande/qubes-gui-agent-linux/commit/23c1eefa89a5c9c299c6f8e66a13b535e7e470d5)

 No. Not that one. This one: https://github.com/QubesOS/qubes-core-agent-linux/commit/a6c95ffd7fcfc1343d62355137688c2c7e6010ef

**Comment by kennethrrosen** [comment]
> cairo software rendering is indeed applied to all in /qubes-sysinit.sh (as patched in #9389) however the problem persists.

yes, as I mentioned above

**Comment by marmarek** [comment]
Better check directly if `GSK_RENDERER` is set to `cairo`. For example in terminal `echo $GSK_RENDERER`. If it is, try opening the application from this very terminal and see if the issue still appears.

**Comment by kennethrrosen** [comment]
```
[user@crypt ~]$ echo $GSK_RENDERER
cairo
[user@crypt ~]$ xdg-open .
[user@crypt ~]$ xdg-open /home/user/QubesIncoming
```
Still occurs. The usage is the same as the initial issue: 

- Open it and try to make duplicated files on different paths lets say /Download /Pictures
- Now try to copy/paste from one path to another
- Result should be shown similar to the screenshot
- Options are missing from the bottom of the window

**Comment by kennethrrosen** [comment]
> > > Is there a way to change a VM's DE from Gnome to XFCE, or vise versa?
> > 
> > 
> > You can use [Xfce templates](https://www.qubes-os.org/doc/templates/xfce/).
> 
> Yes, of course; but in the case of a StandaloneVM?

For the uninitiated and unknowledgeable like me, who may also come here looking for a solution, what I _should_ have asked in the above exchange was how do I switch the file manager. There are plenty to chose from (Thunar and Nautilus being the defaults for xfce and gnome, respectively). I decided to switch to nemo since I couldn't sort out this bug for me:

```
sudo dnf/apt install nemo
xdg-mime default nemo.desktop inode/directory
xdg-mime query default inode/directory # to confirm the vm is now using nemo instead of nautilus/thunar
```

**Comment by ben-grande** [comment]
> EDIT: Note: using another file manager will not include the Qubes copy- move- or sanitization functions

There are integrations for Thunar, Caja, PCMan-ft and Nautilus.

**Comment by ben-grande** [comment]
There is a recommendation from Demi about custom CSS:

- https://github.com/qubesos/qubes-issues/issues/6821

**Comment by ben-grande** [comment]
Tried on Fedora 43, did exactly as described on the first post. No issue. Closing as solved.
